store/tikv/region_cache.go

```diff
@@ -709,8 +766,217 @@ func (r *Region) ContainsByEnd(key []byte) bool {
 		(bytes.Compare(key, r.meta.GetEndKey()) <= 0 || len(r.meta.GetEndKey()) == 0)
 }
 
-// Store contains a tikv server's address.
+// fn loads the Store info given by storeId.
+type resolveFunc func(ctx context.Context, id uint64) (*metapb.Store, error)
+
+// Store contains a kv process's address.
 type Store struct {
-	ID   uint64
-	Addr string
+	addr    atomic.Value // loaded store address(*string)
+	storeID uint64       // store's id
+	state   uint64       // unsafe store storeState
+
+	resolve struct {
+		sync.Mutex             // protect pd from concurrent init requests
+		fn         resolveFunc // func to get store address from PD
+	}
+}
+
+// storeState contains store's access info.
+type storeState struct {
+	lastFailedTime uint32
+	failedAttempt  uint16
+	resolveState   resolveState
+}
```

Comment:
It is too tricky to convert a uint64 to struct and backward... I prefer defining `storeState` as `type storeState uint64` then add methods to it, like `getLastFailedTime()`, `setLastFailedTime` etc. Or simply use `atomic.Value` instead.

Reply:
it's @coocood 's advise...https://github.com/pingcap/tidb/pull/10256#discussion_r280020401

both ways are ok for me, maybe use this unsafe to reduce memory alloc but tricky - -